pyro/infer/mcmc/api.py

```diff
@@ -0,0 +1,322 @@
+"""
+This module offers a modified interface for MCMC inference with the following objectives:
+  - making MCMC independent of Pyro specific trace data structure, to facilitate
+    integration with other PyTorch based libraries.
+  - bringing the interface closer to that of NumPyro to make it easier to write
+    code that works with different backends.
+  - minimal memory consumption with multiprocessing and CUDA.
+"""
+
+from __future__ import absolute_import, division, print_function
+
+import json
+import logging
+import signal
+import threading
+import warnings
+from collections import OrderedDict, defaultdict
+
+import six
+import torch
+import torch.multiprocessing as mp
+from six.moves import queue
+
+import pyro
+from pyro.infer.mcmc.logger import initialize_logger, DIAGNOSTIC_MSG, TqdmHandler, ProgressBar
+
+MAX_SEED = 2**32 - 1
+
+
+def logger_thread(log_queue, warmup_steps, num_samples, num_chains, disable_progbar=False):
+    """
+    Logging thread that asynchronously consumes logging events from `log_queue`,
+    and handles them appropriately.
+    """
+    progress_bars = ProgressBar(warmup_steps, num_samples, disable=disable_progbar, num_bars=num_chains)
+    logger = logging.getLogger(__name__)
+    logger.propagate = False
+    logger.addHandler(TqdmHandler())
+    num_samples = [0] * num_chains
+    try:
+        while True:
+            try:
+                record = log_queue.get(timeout=1)
+            except queue.Empty:
+                continue
+            if record is None:
+                break
+            metadata, msg = record.getMessage().split("]", 1)
+            _, msg_type, logger_id = metadata[1:].split()
+            if msg_type == DIAGNOSTIC_MSG:
+                pbar_pos = int(logger_id.split(":")[-1])
+                num_samples[pbar_pos] += 1
+                if num_samples[pbar_pos] == warmup_steps:
+                    progress_bars.set_description("Sample [{}]".format(pbar_pos + 1), pos=pbar_pos)
+                diagnostics = json.loads(msg, object_pairs_hook=OrderedDict)
+                progress_bars.set_postfix(diagnostics, pos=pbar_pos)
```

Comment:
Maybe we can add the arg `refresh=False` to avoid the flakiness in the speed when using tqdm as in [forum](https://forum.pyro.ai/t/logistic-regression-nuts-on-titanic-dataset-extremely-slow-but-the-time-it-takes-also-wildly-varies/1061/13) (I observed it too).

Reply:
This used to create an issue with the `notebook` env where without the refresh arg, the progress bar wouldn't update (with num_chains > 1). Let me check first if that issue is resolved.